src/imp/propagator_envoy.js

```diff
@@ -0,0 +1,187 @@
+import pb from 'protobufjs';
+import long from 'long';
+import _each from '../_each';
+import _leftpad from '../_leftpad';
+import SpanContextImp from './span_context_imp';
+import LightStepPropagator from './propagator_ls';
+
+const CARRIER_ENVOY_HEADER_KEY = 'x-ot-span-context';
+
+const BINARY_PROTO = {
+    nested : {
+        lightstep : {
+            options : { go_package : 'lightsteppb' },
+            nested  : {
+                BinaryCarrier : {
+                    fields : {
+                        deprecated_text_ctx : {
+                            rule : 'repeated',
+                            type : 'bytes',
+                            id   : 1,
+                        },
+                        basic_ctx : { type : 'BasicTracerCarrier', id : 2 },
+                    },
+                },
+                BasicTracerCarrier : {
+                    fields : {
+                        trace_id      : { type : 'fixed64', id : 1 },
+                        span_id       : { type : 'fixed64', id : 2 },
+                        sampled       : { type : 'bool', id : 3 },
+                        baggage_items : { keyType : 'string', type : 'string', id : 4 },
+                    },
+                },
+            },
+        },
+    },
+};
+
+
+export default class EnvoyPropagator extends LightStepPropagator {
+    constructor(tracer) {
+        super(tracer);
+        this._tracer = tracer;
+        this._envoyHeaderKey = CARRIER_ENVOY_HEADER_KEY;
+        this._carrierPb = pb.Root.fromJSON(BINARY_PROTO);
+    }
+
+    inject(spanContext, carrier) {
+        if (!carrier) {
+            this._tracer._error('Unexpected null carrier in call to inject');
+            return;
+        }
+        if (typeof carrier !== 'object') {
+            this._tracer._error(
+                `Unexpected '${typeof carrier}' FORMAT_BINARY carrier in call to inject`,
+            );
+            return;
+        }
+
+        let basicContext = {
+            span_id       : long.fromString(spanContext._guid, 16),
+            trace_id      : long.fromString(spanContext._traceGUID, 16),
+            sampled       : true,
+            baggage_items : {},
+        };
+        spanContext.forEachBaggageItem((key, value) => {
+            basicContext.baggage_items[key] = value;
+        });
+
+        let binaryCarrier = this._carrierPb.lookupType('BinaryCarrier');
+
+        let payload = {
+            basic_ctx : basicContext,
+        };
+
+        let err = binaryCarrier.verify(payload);
+        if (err) {
+            this._tracer._error(`Invalid Span Context: ${err}`);
+            return null;
+        }
+        let msg = binaryCarrier.create(payload);
+        let buffer = binaryCarrier.encode(msg).finish();
+        let bufferString = pb.util.base64.encode(buffer, 0, buffer.length);
+        carrier[this._envoyHeaderKey] = bufferString;
```

Comment:
I could be mistaken, but for binary, isn't the carrier just a byte array? If so, I don't think there is a notion of a string key, with a binary value. I think it just writes bytes into the array.

Reply:
it still needs a key in the headers, and you still need to base64 the output. maybe this would be one-step with google protos?

keep in mind that adding another proto will cause the tracer file size to creep up again, to your above point

Reply:
Do any of the other Lightstep tracers have an EnvoyPropagator? As I look around I only see BinaryPropagator (this is true for Go and Python anyway).

Reply:
C++ has one too that just uses the binary propagator to do the encoding but then puts it in the [envoy key header.](https://github.com/lightstep/lightstep-tracer-cpp/blob/8f7820883671da285dea697cbb66bf9ebcfc68bd/src/tracer/propagation/envoy_propagator.cpp#L9)

Reply:
I was thinking that we should probably try to do the same. Would it be possible to split to this work into a BinaryPropagator and an Envoy propagator (or instrumentation) that uses it to do the header manipulation?